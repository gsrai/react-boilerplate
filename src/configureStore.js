import { compose, createStore, applyMiddleware } from 'redux'
import createSagaMiddleware, { END } from 'redux-saga'
import rootReducer from './reducers'
import rootSaga from './sagas'

import { composeWithDevTools } from 'redux-devtools-extension'

export default function configureStore (history, initialState) {
  const sagaMiddleware = createSagaMiddleware()
  let store = null

  if (process.env.NODE_ENV === 'production') {
    store = createStore(
      rootReducer,
      initialState,
      compose(
        applyMiddleware(sagaMiddleware)
      )
    )
  } else {
    const DevToolsExt = typeof (window) !== 'undefined' && window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__() // eslint-disable-line

    store = createStore(
      rootReducer,
      initialState,
      // used composeWithDevTools instead of redux's compose
      // in order to have access to the redux-devtools-extension
      composeWithDevTools(
        applyMiddleware(sagaMiddleware)
      )
    )

    // Hot reload reducers (requires Webpack or Browserify HMR to be enabled)
    if (module.hot) {
      module.hot.accept('./reducers', () => {
        const nextRootReducer = require('./reducers').default // eslint-disable-line global-require
        store.replaceReducer(nextRootReducer)
      })
    }
  }

  store.close = () => store.dispatch(END)
  sagaMiddleware.run(rootSaga)
  return store
}
