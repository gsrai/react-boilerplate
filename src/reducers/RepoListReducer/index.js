export const LOAD_REPO_LIST_ACTION_INIT = 'LOAD_REPO_LIST_ACTION_INIT'
export const LOAD_REPO_LIST_ACTION = 'LOAD_REPO_LIST_ACTION'
export const LOAD_REPO_LIST_ACTION_SUCCEDED = 'LOAD_REPO_LIST_ACTION_SUCCEDED'
export const LOAD_REPO_LIST_ACTION_FAILED = 'LOAD_REPO_LIST_ACTION_FAILED'

export const actions = {
  loadRepoListInit: () => ({ type: LOAD_REPO_LIST_ACTION_INIT }),
  loadRepoList: (repoName) => ({ type: LOAD_REPO_LIST_ACTION, query: { q: repoName } }),
  loadRepoListSuccess: (response) => ({ type: LOAD_REPO_LIST_ACTION_SUCCEDED, response }),
  loadRepoListFailure: (reason) => ({ type: LOAD_REPO_LIST_ACTION_FAILED, reason })
}

export const initialState = {
  loading: false,
  loadSuccess: false,
  loadError: false,
  repos: []
}

export default function RepoListReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_REPO_LIST_ACTION_INIT:
      return initialState

    case LOAD_REPO_LIST_ACTION:
      return {
        ...state,
        loading: true,
        loadSuccess: false,
        loadError: false
      }

    case LOAD_REPO_LIST_ACTION_SUCCEDED:
      return {
        ...state,
        loading: false,
        loadSuccess: true,
        loadError: false,
        repos: action.response
      }

    case LOAD_REPO_LIST_ACTION_FAILED:
      return {
        ...state,
        loading: false,
        loadSuccess: false,
        loadError: true,
        reason: action.reason
      }

    default:
      return state
  }
}
