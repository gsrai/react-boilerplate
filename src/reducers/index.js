import { combineReducers } from 'redux'

import repoListReducer from './RepoListReducer'

const rootReducer = combineReducers({
  repoListReducer
})

export default rootReducer
