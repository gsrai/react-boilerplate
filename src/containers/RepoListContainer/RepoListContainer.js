import React from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import { actions as repoListActions } from '../../reducers/RepoListReducer'
import RepoList from '../../components/RepoList'

class RepoListContainer extends React.Component {
  componentDidMount() {
    this.props.actions.loadRepoListInit()
    this.props.actions.loadRepoList('react')
  }

  render() {
    const { loading, loadSuccess, loadError, reason, repos } = this.props
    return (
      <React.Fragment>
        <RepoList
          loading={loading}
          loadSuccess={loadSuccess}
          loadError={loadError}
          reason={reason}
          repos={repos}
        />
      </React.Fragment>
    )
  }
}

RepoListContainer.propTypes = {
  loading: PropTypes.bool.isRequired,
  loadSuccess: PropTypes.bool.isRequired,
  loadError: PropTypes.bool.isRequired,
  reason: PropTypes.string,
  repos: PropTypes.array,
  actions: PropTypes.object.isRequired
}

export default connect(
  state => {
    const loading = state.repoListReducer.loading
    const loadSuccess = state.repoListReducer.loadSuccess
    const loadError = state.repoListReducer.loadError
    const reason = state.repoListReducer.reason

    const repos = state.repoListReducer.repos

    return { loading, loadSuccess, loadError, reason, repos }
  },
  dispatch => ({
    actions: bindActionCreators(repoListActions, dispatch)
  })
)(RepoListContainer)
