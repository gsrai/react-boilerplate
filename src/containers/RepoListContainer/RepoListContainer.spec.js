import { shallow } from 'enzyme'
import React from 'react'

import configureStore from 'redux-mock-store'
import RepoListContainer from './RepoListContainer'
import { actions } from '../../reducers/RepoListReducer'

describe('Component', () => {
  const mockStore = configureStore()
  const store = mockStore({
    repoListReducer: {
      loading: false,
      loadSuccess: false,
      loadError: false,
      repos: [],
      reason: ''
    }
  })

  beforeEach(() => store.clearActions())

  test('component should render connected component', () => {
    const wrapper = shallow(<RepoListContainer store={store} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('component should render correctly', () => {
    const wrapper = shallow(<RepoListContainer store={store} />).dive()
    expect(wrapper).toMatchSnapshot()
  })

  test('component should dispatch init and load actions on mount', () => {
    shallow(<RepoListContainer store={store} />).dive()
    const dispatchedActions = store.getActions()
    expect(dispatchedActions).toEqual([
      actions.loadRepoListInit(),
      actions.loadRepoList('react')
    ])
  })
})
