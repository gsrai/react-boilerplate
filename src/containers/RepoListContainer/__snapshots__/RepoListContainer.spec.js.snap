// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component component should render connected component 1`] = `
<RepoListContainer
  actions={
    Object {
      "loadRepoList": [Function],
      "loadRepoListFailure": [Function],
      "loadRepoListInit": [Function],
      "loadRepoListSuccess": [Function],
    }
  }
  loadError={false}
  loadSuccess={false}
  loading={false}
  reason=""
  repos={Array []}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`Component component should render correctly 1`] = `
<Fragment>
  <RepoList
    loadError={false}
    loadSuccess={false}
    loading={false}
    reason=""
    repos={Array []}
  />
</Fragment>
`;
